### ISSUE TITLE
newbie question: slow training on colab possible due to giant training set

### ISSUE THREAD

**Issue by sloev** [amaiya/ktrain#16]
First of all: **thank you** this lib makes getting your hands dirty much faster.

second:

I managed to get everything running in colab with a dataset i've created for doing sentiment analysis. I have quite a lot of data so i thought "why not throw a lot at BERT" then:

```
begin training using triangular learning rate policy with max lr of 2e-05...
Train on 195000 samples, validate on 76263 samples
Epoch 1/3
   416/195000 [..............................] - ETA: 51:10:26 - loss: 1.6178 - acc: 0.2404
```

As you can see i am training on 195'000 items and validating on 76263 items and the grand `question` now is **Thats too much data right?** cause i dont like the ETA up there: 51 hours!

And that is 51 hours on google colab with GPU support etc.

**Comment by amaiya** [comment]
BERT is quite slow, since it is a large model with lots of parameters (see [here](https://blog.rasa.com/compressing-bert-for-faster-prediction-2/)).   Some options to address this:
1. It's possible to train BERT using a TPU instead of GPU (but, this not yet supported directly through *ktrain*).  See [here](https://www.dlology.com/blog/how-to-train-keras-model-x20-times-faster-with-tpu-for-free/) for more information.  This requires using tf.keras instead of standalone Keras.  tf.keras will be fully supported in *ktrain* in the near future (experimental support exists now, but it is not well-tested). 
2. Use ALBERT, which is a leaner version of BERT that was just released.  This isn't supported in *ktrain* yet, but maybe in the future
3. Use a smaller faster model in *ktrain*.   For text classification datasets, you'll sometimes find that the BERT performance boost is only about 1 to 3 percentage points in accuracy at most.   So, maybe you can live with a slightly lower accuracy of a model that trains fast? 
4. You can try freezing some of the earlier layers (see `learner.freeze` method).  Note that I've only used this with image classification models and this has not been tried/tested specifically with BERT.

Like I said, option 1 and 2 is not available in *ktrain* right now, but probably will be in the future.

Another question you might ask yourself is whether or not you really need a training set that large.  Part of the power of pre-trained models like BERT and pre-trained ImageNet models in computer vision is that you can get away with using smaller training sets when fine-tuning.  If you train on a smaller sample and validate on the rest and get good results, maybe you don't need to train for that long.

**Comment by sloev** [comment]
Thank you @amaiya for a great and thorough answer. i will look into your proposals and i thank you for your time

**Comment by agn-7** [comment]
In addition to @amaiya  brilliant tips, you could decrease your model train time by increasing the `batch_size` to whatever colab's RAM capacity supports. For example, I use `batch_sieze=128` instead of default `ktrain.text.bert` with `batch_size=6`:

```learner = ktrain.get_learner(model, train_data=(x_train, y_train), batch_size=128)```

Another point which is really efficient is set GPU as runtime in colab instead of default mode (NONE/CPU). In fact, it speeds up your train approximately 8x~50x faster. Note that, you have to restart the colab runtime in order to apply that. However, I think you aware of that and you have activated it.
